File: python/112_path-sum.py
from collections import deque

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# DFS with Recursion
class Solution:
    def hasPathSum(self, root: TreeNode, Sum: int) -> bool:
        if not root:
            return

        if not root.left and not root.right and Sum == root.val:
            return True

        return self.hasPathSum(root.left, Sum - root.val) or self.hasPathSum(root.right, Sum - root.val)


#---------------------------------------------------------------------------------------#


# DFS with Stack
class Solution:
    def hasPathSum(self, root: TreeNode, targetSum: int) -> bool:
        if not root:
            return False

        stack = deque()
        stack.append((root, root.val))

        while len(stack) > 0:
            node, currentSum = stack.pop()

            if not node.left and not node.right and currentSum == targetSum:
                return True

            if node.left:
                stack.append((node.left, currentSum + node.left.val))
            if node.right:
                stack.append((node.right, currentSum + node.right.val))

        return False

File: go/112_path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// DFS with Recursion
func hasPathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}

	if root.Left == nil && root.Right == nil {
		return root.Val == targetSum
	}

	return hasPathSum(root.Left, targetSum-root.Val) || hasPathSum(root.Right, targetSum-root.Val)
}

/*---------------------------------------------------------------------------------------*/

// BFS with Queue
type queue struct {
	node *TreeNode
	sum  int
}

func hasPathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}

	var q []queue
	q = append(q, queue{root, root.Val})

	for len(q) > 0 {
		node, sum := q[0].node, q[0].sum
		q = q[1:]

		if node.Left == nil && node.Right == nil {
			if sum == targetSum {
				return true
			}
		}

		if node.Left != nil {
			q = append(q, queue{node.Left, sum + node.Left.Val})
		}
		if node.Right != nil {
			q = append(q, queue{node.Right, sum + node.Right.Val})
		}
	}
	return false
}